Move non_blocking copies to aten

li-roy opened the pull request:
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13603 Move THTensor_(copy) to aten&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D12936031/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13348 Move cuda copy to aten&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D12850690/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#13866 Move non_blocking copies to aten**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13011878/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13937 remove copy_wrapper&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13053648/)

just a straightforward port


Differential Revision: [D13011878](https://our.internmc.facebook.com/intern/diff/D13011878/)

Commit:
V1: Initial commit

Differential Revision: D13011878
Differential Version: 63368756

Commit:
V2: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63372863

Commit:
V3: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63464543

Commit:
V4: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63495321

Commit:
V5: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63508465

Commit:
V7: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63718803

Commit:
V8: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63733454

Commit:
V9: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63768505

Commit:
V10: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 63837642

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
   });
 }
 
+void copy_from_cpu_async_(Tensor& dst, const Tensor& src) {
```
THCTensor_(copyAsyncCPU)

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
 
+void copy_from_cpu_async_(Tensor& dst, const Tensor& src) {
+  AT_ASSERT(dst.is_contiguous());
+  AT_ASSERT(src.is_contiguous());
```
You seem to have lost the check that dst and src have same numel?

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
 }
 
+void copy_from_cpu_async_(Tensor& dst, const Tensor& src) {
+  AT_ASSERT(dst.is_contiguous());
```
These probably shouldn't be asserts; I'm not 100% it's possible to trigger this error condition from user code, but the old TH code seemed to actually give user visible error messages in this case. `AT_ASSERT` means "We, the PyTorch devs, did something wrong, please yell at us."

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+    return;
+  }
+
+  DeviceGuard device_guard{dst.device()};
```
Use a CUDADeviceGuard here; no reason not to.

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+  DeviceGuard device_guard{dst.device()};
+  CUDAStream stream = getCurrentCUDAStream();
+
+  AT_DISPATCH_ALL_TYPES_AND_HALF(src.type(), "copy_from_cpu_async", [&]() {
```
I don't think there's any reason to do a dispatch here. You can use void* pointers, and compute the size of the scalar type using itemsize()

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+        cudaMemcpyHostToDevice,
+        stream));
+    AT_CUDA_CHECK(THCCachingHostAllocator_recordEvent(
+        src.storage().data<scalar_t>(), stream.internals()));
```
I have a patch coming down the pipe that will conflict with this. We'll see who lands first :)

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+  });
+}
+
+void copy_to_cpu_async_(Tensor& dst, const Tensor& src) {
```
THTensor_(copyAsyncCuda)

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+
+void copy_to_cpu_async_(Tensor& dst, const Tensor& src) {
+  AT_ASSERT(dst.is_contiguous());
+  AT_ASSERT(src.is_contiguous());
```
Same deal as above; missing a check, and using the wrong AT_ macro

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+    return;
+  }
+
+  DeviceGuard device_guard{dst.device()};
```
This is wrong, it should be src.device()

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+  DeviceGuard device_guard{dst.device()};
+  CUDAStream stream = getCurrentCUDAStream();
+
+  AT_DISPATCH_ALL_TYPES_AND_HALF(src.type(), "copy_to_cpu_async", [&]() {
```
Yeah, I think we don't need a dispatch here either.

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
       if (std::is_same<dst_T, scalar_t>::value) {
+        if (non_blocking) {
+          copy_from_cpu_async_(dst, src);
+          return;
```
nit: I think I'd prefer it if this code was written without the short-circuiting returns. I don't think it will increase the indentation level too much.

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -200,23 +200,79 @@ void copy_to_cpu(Tensor& dst, const Tensor& src) {
+          return;
+        }
         copy_to_cpu(dst, src);
+        return;
```
How did this ever work before? ;)

ezyang reviewed (changes requested):
Still seem to be some mistakes. CI didn't flag them, so I wonder if we're actually exercising these codepaths... intruiging...

Commit:
V11: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 64537331

ezyang reviewed (approved):
Nice update. What's your plan for increasing test coverage?

Commit:
V12: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 64889682

li-roy commented:
It was actually working, but in a really weird and unintended way. Added an assert to CUDAGuard so it can't be created with the cpu device type so this doesn't happen anymore. Also added a couple tests.

Commit:
V13: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 64905030

Commit:
V14: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 64917224

ezyang commented:
Is the assert in this diff, or a different one?

Commit:
V15: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 64937521

Commit:
V16: Merge with parent diff changes

Differential Revision: D13011878
Differential Version: 64996418
